File: android/app/src/main/kotlin/com/roomcheck/app/data/NightStore.kt
package com.roomcheck.app.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

enum class Mark { IN, OUT, EXC }

data class Night(
    val marks: MutableMap<String, MutableMap<String, Mark>> = mutableMapOf(),
    val excusedTonight: MutableSet<String> = mutableSetOf(),
    val notes: MutableMap<String, String> = mutableMapOf(),
    var closed: Boolean = false,
    /** When each cell was last set, so two devices' copies can be merged cell by cell. */
    val stamps: MutableMap<String, Long> = mutableMapOf()
) {
    /** Records that a cell just changed. Every edit goes through here or it will not sync. */
    fun touch(key: String, at: Long = System.currentTimeMillis()) { stamps[key] = at }

    fun toJson(): JSONObject {
        val o = JSONObject()
        val marksObj = JSONObject()
        // the night's own slots, not a fixed three: a custom time has to survive being saved
        marks.forEach { (sid, slot) ->
            val obj = JSONObject()
            slot.forEach { (pid, mark) -> obj.put(pid, mark.name.lowercase()) }
            marksObj.put(sid, obj)
        }
        o.put("marks", marksObj)
        o.put("tonight", JSONArray(excusedTonight.toList()))
        val notesObj = JSONObject()
        notes.forEach { (pid, note) -> notesObj.put(pid, note) }
        o.put("notes", notesObj)
        o.put("closed", closed)
        val stampObj = JSONObject()
        stamps.forEach { (k, v) -> stampObj.put(k, v) }
        o.put("ts", stampObj)
        return o
    }

    companion object {
        /**
         * [fallbackStamp] dates cells in a night saved before timestamps existed - the file's own
         * mtime, which is roughly when those marks were made. Without it they would all read as
         * "never set" and the first sync from any other device would wipe them.
         */
        fun fromJson(json: JSONObject?, fallbackStamp: Long = 0L): Night {
            val n = Night()
            if (json == null) return n
            val marksObj = json.optJSONObject("marks")
            marksObj?.keys()?.forEach { sid ->
                val slot = marksObj.optJSONObject(sid) ?: return@forEach
                val map = mutableMapOf<String, Mark>()
                slot.keys().forEach { pid ->
                    when (slot.optString(pid)) {
                        "in" -> map[pid] = Mark.IN
                        "out" -> map[pid] = Mark.OUT
                        "exc" -> map[pid] = Mark.EXC
                    }
                }
                n.marks[sid] = map
            }
            val tonight = json.optJSONArray("tonight")
            if (tonight != null) for (i in 0 until tonight.length()) n.excusedTonight.add(tonight.getString(i))
            val notesObj = json.optJSONObject("notes")
            notesObj?.keys()?.forEach { pid -> n.notes[pid] = notesObj.optString(pid) }
            n.closed = json.optBoolean("closed", false)

            val stampObj = json.optJSONObject("ts")
            if (stampObj != null) {
                stampObj.keys().forEach { k -> n.stamps[k] = stampObj.optLong(k) }
            } else if (fallbackStamp > 0L) {
                n.marks.forEach { (sid, slot) -> slot.keys.forEach { n.stamps[Merge.markKey(sid, it)] = fallbackStamp } }
                n.excusedTonight.forEach { n.stamps[Merge.tonightKey(it)] = fallbackStamp }
                n.notes.keys.forEach { n.stamps[Merge.noteKey(it)] = fallbackStamp }
                if (n.closed) n.stamps[Merge.CLOSED_KEY] = fallbackStamp
            }
            return n
        }
    }
}

data class PersonOverride(var first: String? = null, var last: String? = null, var always: Boolean = false, var reason: String? = null)

/**
 * The things you can turn on and off, kept off by default: the plan and the sheet read cleaner
 * without them, and anyone who wants them back can say so in Names.
 */
data class Settings(
    /** "top"/"bottom" written on each half of a bunk. */
    val bunkLabels: Boolean = false,
    /** Hebrew names on the floor plan you tap through. */
    val hebrewOnPlan: Boolean = false,
    /** Hebrew names in the sent picture. Separate on purpose: you may want one and not the other. */
    val hebrewInExport: Boolean = false,
    /** Where the shared copy lives. Blank means this device keeps to itself, as it always did. */
    val syncUrl: String = "",
    val syncToken: String = ""
)

private fun JSONObject.optStringOrNull(key: String): String? = if (has(key)) getString(key) else null

/**
 * Every night lives in a file under [home] - on the phone that is the app's private storage, which
 * is why the data is the app's own and not the browser's. [home] is a plain directory rather than a
 * Context so the screenshot tests can point a real store at a scratch folder and render the actual
 * screen instead of a stand-in.
 */
class NightStore(home: File) {
    constructor(context: Context) : this(context.filesDir)

    private val dir = File(home, "nights").apply { mkdirs() }
    private val stateFile = File(home, "state.json")

    var extra: MutableMap<String, PersonOverride> = mutableMapOf()
        private set

    var settings: Settings = Settings()

    /** How far through the server's changes this device has read. */
    var lastPull: Long = 0L

    // Nights edited here since the last successful push. Kept on disk because the edit may have
    // happened on a bus with no signal and the app may be killed before it ever gets one.
    private val dirty = mutableSetOf<String>()

    fun dirtyDates(): Set<String> = dirty.toSet()
    fun clearDirty(done: Set<String>) { dirty.removeAll(done); saveState() }

    init { loadState() }

    private fun loadState() {
        extra = mutableMapOf()
        if (stateFile.exists()) {
            runCatching {
                val root = JSONObject(stateFile.readText())
                root.optJSONObject("settings")?.let {
                    settings = Settings(
                        bunkLabels = it.optBoolean("bunkLabels", false),
                        hebrewOnPlan = it.optBoolean("hebrewOnPlan", false),
                        hebrewInExport = it.optBoolean("hebrewInExport", false),
                        syncUrl = it.optString("syncUrl", ""),
                        syncToken = it.optString("syncToken", "")
                    )
                }
                lastPull = root.optLong("lastPull", 0L)
                root.optJSONArray("dirty")?.let { arr ->
                    for (i in 0 until arr.length()) dirty.add(arr.getString(i))
                }
                val extraObj = root.optJSONObject("extra")
                extraObj?.keys()?.forEach { pid ->
                    val po = extraObj.getJSONObject(pid)
                    extra[pid] = PersonOverride(
                        first = po.optStringOrNull("first"),
                        last = po.optStringOrNull("last"),
                        always = po.optBoolean("always", false),
                        reason = po.optStringOrNull("reason")
                    )
                }
            }
        }
    }

    fun saveState() {
        val root = JSONObject()
        val extraObj = JSONObject()
        extra.forEach { (pid, po) ->
            val o = JSONObject()
            po.first?.let { o.put("first", it) }
            po.last?.let { o.put("last", it) }
            o.put("always", po.always)
            po.reason?.let { o.put("reason", it) }
            extraObj.put(pid, o)
        }
        root.put("extra", extraObj)
        root.put("settings", JSONObject()
            .put("bunkLabels", settings.bunkLabels)
            .put("hebrewOnPlan", settings.hebrewOnPlan)
            .put("hebrewInExport", settings.hebrewInExport)
            .put("syncUrl", settings.syncUrl)
            .put("syncToken", settings.syncToken))
        root.put("lastPull", lastPull)
        root.put("dirty", JSONArray(dirty.toList()))
        stateFile.writeText(root.toString())
    }

    private fun fileFor(dateKey: String) = File(dir, "$dateKey.json")

    fun load(dateKey: String): Night {
        val f = fileFor(dateKey)
        if (!f.exists()) return Night()
        // a night written before timestamps existed is dated by its file, not by zero
        return runCatching { Night.fromJson(JSONObject(f.readText()), f.lastModified()) }.getOrElse { Night() }
    }

    fun save(dateKey: String, night: Night) {
        fileFor(dateKey).writeText(night.toJson().toString())
    }

    /** Saves a local edit and remembers the night still owes the server a push. */
    fun saveLocal(dateKey: String, night: Night) {
        save(dateKey, night)
        if (dirty.add(dateKey)) saveState()
    }

    fun hasData(dateKey: String): Boolean = fileFor(dateKey).exists()

    fun allDateKeys(): Set<String> = dir.listFiles()?.mapNotNull {
        it.name.takeIf { n -> n.endsWith(".json") }?.removeSuffix(".json")
    }?.toSet() ?: emptySet()

    // ---- backup / restore, same schema as the web app's dl/doimp handlers ----
    fun exportBackup(): String {
        val root = JSONObject()
        root.put("v", 1)
        root.put("saved", java.time.Instant.now().toString())
        val extraObj = JSONObject()
        extra.forEach { (pid, po) ->
            val o = JSONObject()
            po.first?.let { o.put("first", it) }
            po.last?.let { o.put("last", it) }
            o.put("always", po.always)
            po.reason?.let { o.put("reason", it) }
            extraObj.put(pid, o)
        }
        root.put("extra", extraObj)
        val nights = JSONObject()
        allDateKeys().forEach { key -> nights.put("dn:$key", load(key).toJson().toString()) }
        root.put("nights", nights)
        return root.toString(1)
    }

    fun importBackup(text: String): Boolean {
        return runCatching {
            val root = JSONObject(text)
            val hasNights = root.has("nights")
            val hasExtra = root.has("extra")
            if (!hasNights && !hasExtra) return false

            extra = mutableMapOf()
            root.optJSONObject("extra")?.let { extraObj ->
                extraObj.keys().forEach { pid ->
                    val po = extraObj.getJSONObject(pid)
                    extra[pid] = PersonOverride(
                        first = po.optStringOrNull("first"),
                        last = po.optStringOrNull("last"),
                        always = po.optBoolean("always", false),
                        reason = po.optStringOrNull("reason")
                    )
                }
            }
            saveState()

            root.optJSONObject("nights")?.let { nightsObj ->
                nightsObj.keys().forEach { key ->
                    val dateKey = key.removePrefix("dn:")
                    val nightJsonStr = nightsObj.getString(key)
                    val night = Night.fromJson(JSONObject(nightJsonStr))
                    save(dateKey, night)
                }
            }
            true
        }.getOrElse { false }
    }
}

File: ios/RoomCheckCore/Sources/RoomCheckCore/NightStore.swift
import Foundation

/// Every night lives in a file under [home]. [home] is a plain directory rather than an app's
/// documents URL directly, so tests can point a real store at a scratch folder.
public final class NightStore {
    private let dir: URL
    private let stateFile: URL

    public var extra: [String: PersonOverride] = [:]
    public var settings: Settings = Settings()
    /// How far through the server's changes this device has read.
    public var lastPull: Int64 = 0

    // Nights edited here since the last successful push. Kept on disk because the edit may have
    // happened on a bus with no signal and the app may be killed before it ever gets one.
    private var dirty: Set<String> = []

    public init(home: URL) {
        dir = home.appendingPathComponent("nights", isDirectory: true)
        stateFile = home.appendingPathComponent("state.json")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        loadState()
    }

    public func dirtyDates() -> Set<String> { dirty }
    public func clearDirty(_ done: Set<String>) { dirty.subtract(done); saveState() }

    private func loadState() {
        extra = [:]
        guard let data = try? Data(contentsOf: stateFile),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }

        if let s = root["settings"] as? [String: Any] {
            settings = Settings(
                bunkLabels: (s["bunkLabels"] as? Bool) ?? false,
                hebrewOnPlan: (s["hebrewOnPlan"] as? Bool) ?? false,
                hebrewInExport: (s["hebrewInExport"] as? Bool) ?? false,
                syncUrl: (s["syncUrl"] as? String) ?? "",
                syncToken: (s["syncToken"] as? String) ?? ""
            )
        }
        if let n = root["lastPull"] as? NSNumber { lastPull = n.int64Value }
        if let d = root["dirty"] as? [String] { dirty = Set(d) }
        if let extraObj = root["extra"] as? [String: Any] {
            for (pid, raw) in extraObj {
                guard let po = raw as? [String: Any] else { continue }
                extra[pid] = PersonOverride(
                    first: po["first"] as? String, last: po["last"] as? String,
                    always: (po["always"] as? Bool) ?? false, reason: po["reason"] as? String
                )
            }
        }
    }

    private func extraJSONObject() -> [String: Any] {
        var out: [String: Any] = [:]
        for (pid, po) in extra {
            var o: [String: Any] = ["always": po.always]
            if let v = po.first { o["first"] = v }
            if let v = po.last { o["last"] = v }
            if let v = po.reason { o["reason"] = v }
            out[pid] = o
        }
        return out
    }

    public func saveState() {
        let root: [String: Any] = [
            "extra": extraJSONObject(),
            "settings": [
                "bunkLabels": settings.bunkLabels, "hebrewOnPlan": settings.hebrewOnPlan,
                "hebrewInExport": settings.hebrewInExport, "syncUrl": settings.syncUrl, "syncToken": settings.syncToken
            ],
            "lastPull": lastPull,
            "dirty": Array(dirty)
        ]
        if let data = try? JSONSerialization.data(withJSONObject: root) {
            try? data.write(to: stateFile)
        }
    }

    private func fileFor(_ dateKey: String) -> URL { dir.appendingPathComponent("\(dateKey).json") }

    public func load(_ dateKey: String) -> Night {
        let f = fileFor(dateKey)
        guard let data = try? Data(contentsOf: f) else { return Night() }
        // a night written before timestamps existed is dated by its file, not by zero
        let mtime = (try? FileManager.default.attributesOfItem(atPath: f.path)[.modificationDate] as? Date) ?? nil
        let fallback = Int64(((mtime ?? Date()) as Date).timeIntervalSince1970 * 1000)
        return Night.fromJSONData(data, fallbackStamp: fallback)
    }

    public func save(_ dateKey: String, _ night: Night) {
        if let data = try? night.toJSONData() { try? data.write(to: fileFor(dateKey)) }
    }

    /// Saves a local edit and remembers the night still owes the server a push.
    public func saveLocal(_ dateKey: String, _ night: Night) {
        save(dateKey, night)
        if dirty.insert(dateKey).inserted { saveState() }
    }

    public func hasData(_ dateKey: String) -> Bool { FileManager.default.fileExists(atPath: fileFor(dateKey).path) }

    public func allDateKeys() -> Set<String> {
        let names = (try? FileManager.default.contentsOfDirectory(atPath: dir.path)) ?? []
        return Set(names.filter { $0.hasSuffix(".json") }.map { String($0.dropLast(5)) })
    }

    // MARK: - backup / restore, same schema as the website and the Android app

    public func exportBackup() -> String {
        let root: [String: Any] = [
            "v": 1,
            "saved": ISO8601DateFormatter().string(from: Date()),
            "extra": extraJSONObject(),
            "nights": Dictionary(uniqueKeysWithValues: allDateKeys().map { key in
                ("dn:\(key)", String(data: (try? load(key).toJSONData()) ?? Data(), encoding: .utf8) ?? "{}")
            })
        ]
        let data = (try? JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted, .sortedKeys])) ?? Data()
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    public func importBackup(_ text: String) -> Bool {
        guard let data = text.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return false }
        let hasNights = root["nights"] != nil
        let hasExtra = root["extra"] != nil
        guard hasNights || hasExtra else { return false }

        extra = [:]
        if let extraObj = root["extra"] as? [String: Any] {
            for (pid, raw) in extraObj {
                guard let po = raw as? [String: Any] else { continue }
                extra[pid] = PersonOverride(
                    first: po["first"] as? String, last: po["last"] as? String,
                    always: (po["always"] as? Bool) ?? false, reason: po["reason"] as? String
                )
            }
        }
        saveState()

        if let nightsObj = root["nights"] as? [String: Any] {
            for (key, raw) in nightsObj {
                guard let jsonStr = raw as? String, let jsonData = jsonStr.data(using: .utf8) else { continue }
                let dateKey = key.hasPrefix("dn:") ? String(key.dropFirst(3)) : key
                save(dateKey, Night.fromJSONData(jsonData))
            }
        }
        return true
    }
}
